secp256kfun/src/poly.rs

```diff
@@ -0,0 +1,323 @@
+//! Utilities for working with polynomials on the secp256k1 elliptic curve.
+//!
+//! A polynomial f(x) of degree k is defined by its coefficients
+//!
+//! `f(x) = a_0 + a_1 * x + a_2 * x^2 + ... + a_k * x^k`
+use core::iter;
+
+#[cfg(feature = "alloc")]
+use alloc::vec::Vec;
+use rand_core::RngCore;
+
+use crate::{
+    g,
+    marker::{NonNormal, PointType, Public, Secrecy, Secret, Zero, ZeroChoice},
+    s, Point, Scalar, G,
+};
+
+/// Create a vector of points by multiplying each scalar by `G`.
+///
+/// # Example
+///
+/// ```
+/// use secp256kfun::{g, poly, s, Scalar, G};
+/// let secret_poly = (0..5)
+///     .map(|_| Scalar::random(&mut rand::thread_rng()))
+///     .collect::<Vec<_>>();
+/// let point_poly = poly::to_point_poly(&secret_poly);
+/// ```
+pub fn to_point_poly(scalar_poly: &[Scalar]) -> Vec<Point> {
+    scalar_poly.iter().map(|a| g!(a * G).normalize()).collect()
+}
+
+/// Generate a [`Scalar`] polynomial for key generation
+///
+/// [`Scalar`]: crate::Scalar
+pub fn generate_scalar_poly(threshold: usize, rng: &mut impl RngCore) -> Vec<Scalar> {
+    (0..threshold).map(|_| Scalar::random(rng)).collect()
+}
+
+/// Evaluate a scalar polynomial defined by coefficients, at some scalar index.
+///
+/// The polynomial coefficients begin with the smallest degree term first (the constant).
+pub fn scalar_poly_eval(
+    poly: &[Scalar],
+    x: Scalar<impl Secrecy, impl ZeroChoice>,
+) -> Scalar<Secret, Zero> {
+    s!(powers(x) .* poly)
+}
+
+/// Evaluate a point polynomial defined by coefficients, at some index.
+///
+/// The polynomial coefficients begin with the smallest degree term first (the constant).
+pub fn point_poly_eval(
```

Comment:
I still think it's better separate operations on `&[Scalar]` and `&[Point]` in different modules. It can be in this file just put `mod point { ... }` around the stuff and remove `point` from the function names (and do the same for scalars).

Reply:
Also adding `interpolate_point_polynomial` to `poly::point::interpolate`